### PR TITLE
Updating Duo SDK to v4 with Universal Prompt support

### PR DESCRIPTION
**Pull request by gershom96**
## Purpose
The purpose of this PR is to update the Duo Authenticator version to sdk v4.

### CODE REVIEW

File: component/authenticator/pom.xml [hunk 1]
```diff
@@ -23,7 +23,7 @@
         <relativePath>../../pom.xml</relativePath>
     </parent>
     <artifactId>org.wso2.carbon.extension.identity.authenticator.duo.connector</artifactId>
-    <version>1.1.18-SNAPSHOT</version>
+    <version>1.1.17</version>
```
Comment (omindu) [inline component/authenticator/pom.xml]: What's the reason for changing this ?

Comment (gershom96) [reply component/authenticator/pom.xml]: resolved with 76916f91a28ab923f57fa4ffa05093fd5c150b24

File: component/authenticator/pom.xml [hunk 2]
```diff
@@ -54,6 +54,31 @@
...
         </dependency>
+        <dependency>
+            <groupId>com.squareup.retrofit2</groupId>
+            <artifactId>retrofit</artifactId>
+            <version>2.9.0</version>
+        </dependency>
+        <dependency>
+            <groupId>com.squareup.retrofit2</groupId>
+            <artifactId>converter-jackson</artifactId>
+            <version>2.9.0</version>
```
Comment (omindu) [inline component/authenticator/pom.xml]: move all the versions as properties in parent pom

File: component/authenticator/pom.xml [hunk 3]
```diff
@@ -190,8 +228,21 @@
                             javax.servlet.http,
                             org.wso2.carbon.extension.identity.helper.*;
                             version="${identity.extension.utils.import.version.range}",
+                            org.wso2.carbon.identity.core.*,
+                            org.wso2.carbon.identity.core*,
```
Comment (omindu) [inline component/authenticator/pom.xml]: What's the requirement to add 2 flavors of this import ?

File: component/authenticator/pom.xml (hunk 2 above)
Comment (gershom96) [reply component/authenticator/pom.xml]: resolved

File: component/authenticator/pom.xml [hunk 4]
```diff
@@ -263,8 +268,8 @@
         </plugins>
     </build>
     <properties>
-        <maven.checkstyleplugin.version>3.1.0</maven.checkstyleplugin.version>
-        <maven.findbugsplugin.version>3.0.5</maven.findbugsplugin.version>
-        <findbugs.annotations.version>3.0.1</findbugs.annotations.version>
+<!--        <maven.checkstyleplugin.version>3.1.0</maven.checkstyleplugin.version>-->
```
Comment (omindu) [inline component/authenticator/pom.xml]: do we need this ?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/duo/DuoAuthenticator.java [hunk 5]
```diff
@@ -469,28 +537,51 @@ public List<Property> getConfigurationProperties() {
...
+
+            String username = authenticatedUser.getAuthenticatedSubjectIdentifier();
+
+            if (username == null) {
+                throw new AuthenticationFailedException("Authentication failed!. Cannot proceed further without " +
+                        "identifying the user");
+            }
+            // Step 6: Exchange the auth duoCode for a Token object which contains metadata about authentication.
+            String duoUserId = getDuoUserId(context);
+            Token duoToken = duoClient.exchangeAuthorizationCodeFor2FAResult(requestDuoCode, duoUserId);
```
Comment (gershom96) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/duo/DuoAuthenticator.java:591]: we have to either persist the duoClient or create a new duoClient and get the token. Give preference to creating a new client.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/duo/DuoAuthenticator.java [hunk 6]
```diff
@@ -396,6 +391,77 @@ private String getMobileNumberForFederatedUser(Map<ClaimMapping, String> userAtt
...
+    }
+
+    /**
+     * Extract the mobile number value from federated user attributes.
+     *
+     * @param context         the AuthenticationContext
+     */
+    private String getDuoUserId(AuthenticationContext context) throws AuthenticationFailedException {
+
+        log.debug("Read the Claim value which was set during the  will be passed to duo.");
```
Comment (gershom96) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/duo/DuoAuthenticator.java]: add this to an if block

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/duo/DuoAuthenticator.java [hunk 7]
```diff
@@ -396,6 +391,77 @@ private String getMobileNumberForFederatedUser(Map<ClaimMapping, String> userAtt
...
+
+        Map<String, String> runtimeParams = getRuntimeParams(context);
+
+        if (runtimeParams != null) {
+            String duoUserId = runtimeParams.get(DuoAuthenticatorConstants.DUO_USER_IDENTIFIER);
+            if (StringUtils.isNotBlank(duoUserId)) {
+                log.debug("The claim obtained from the runtime parameters was passed to Duo.");
```
Comment (gershom96) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/duo/DuoAuthenticator.java]: check all log.debug and add a log.isDebugEnbaled()

**Review by sahandilshan — APPROVED** [review]

**Pull request merged**